### PR TITLE
INT-3965: AMQP Header Mapping Changes

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3965

### CODE REVIEW

**Comment by artembilan** [comment]
Looks like this PR has dependency on the https://github.com/spring-projects/spring-amqp/pull/387.
That one has been just merged.
So, will see how this works on the fresh SA version.

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java [hunk 1]
```diff
@@ -405,4 +411,50 @@ private void addConsumerMetadata(MessageProperties messageProperties, Map<String
...
+
+	/**
+	 * @return the default reply headers for an outbound mapper.
+	 */
+	public static String[] outboundReplyHeaders() {
+		return safeInboundHeaders();
+	}
+
+	private static String[] safeInboundHeaders() {
+		return new String[] { "!x-*", "*" };
```
Comment (artembilan) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java]: Cool! 
Negate in the action!
:+1: :smile:

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java [hunk 2]
```diff
@@ -405,4 +411,50 @@ private void addConsumerMetadata(MessageProperties messageProperties, Map<String
...
+	 */
+	public static String[] outboundRequestHeaders() {
+		return new String[] { "*" };
+	}
+
+	/**
+	 * @return the default reply headers for an outbound mapper.
+	 */
+	public static String[] outboundReplyHeaders() {
+		return safeInboundHeaders();
```
Comment (artembilan) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java]: Is it intentional for `reply` part?
Just for sure.

File: src/reference/asciidoc/amqp.adoc [hunk 3]
```diff
@@ -1058,23 +1058,32 @@ public IntegrationFlow pubSubInFlow(ConnectionFactory connectionFactory) {
...
-Of course, you can pass in your own implementation of AMQP specific header mappers, as the adapters have respective properties to support that.
+Of course, you can pass in your own implementation of AMQP specific header mappers, as the adapters have respective
+properties to support that.
 
-Any user-defined headers within the AMQP http://docs.spring.io/spring-amqp/api/org/springframework/amqp/core/MessageProperties.html[MessageProperties] will NOT
-be copied to or from an AMQP Message, unless explicitly specified by the _requestHeaderNames_ and/or
+Any user-defined headers within the AMQP http://docs.spring.io/spring-amqp/api/org/springframework/amqp/core/MessageProperties.html[MessageProperties] WILL
+be copied to or from an AMQP Message, unless explicitly negated by the _requestHeaderNames_ and/or
 _replyHeaderNames_ properties of the `DefaultAmqpHeaderMapper`.
+For an inbound mapper, the deliveryMode and all `x-*` headers are not mapped by default.
```
Comment (artembilan) [inline src/reference/asciidoc/amqp.adoc]: I'd make deliveryMode as a `deliveryMode`.
Just for better readability.
But it is minor and I don't see anything to complain about.
So, I'm fine to merge after your answer to that my question regarding the reply mapping for inbound adapter.

**Comment by artembilan** [comment]
A general comment: do we consider to do something similar for all other `AbstractHeaderMapper` implementors? WS, XMPP.
Separate JIRA, of course. Just in case to know your opinion.

File: spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java (hunk 2 above)
Comment (garyrussell) [inline spring-integration-amqp/src/main/java/org/springframework/integration/amqp/support/DefaultAmqpHeaderMapper.java]: Yes, consider `ibgw -> obgw`; without this, the reply headers will be propagated.

File: src/reference/asciidoc/amqp.adoc (hunk 3 above)
Comment (garyrussell) [inline src/reference/asciidoc/amqp.adoc]: The `deliveryMode` comment can now be removed; that was from before AMQP pull 387

> For an inbound mapper, `x-*` headers are not mapped by default.

**Comment by garyrussell** [comment]
I would say let's wait until someone asks for it - I am not so sure custom headers are so common there.

File: src/reference/asciidoc/amqp.adoc (hunk 3 above)
Comment (artembilan) [inline src/reference/asciidoc/amqp.adoc]: OK. Thank you for explanation.
Merging with this minor fix.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/78aaa6dac613762455073e45f1ef30ad262e56a7

**Pull request closed without merging**